--- a/userspace/private/libs/wlcsm/Makefile
+++ b/userspace/private/libs/wlcsm/Makefile
@@ -73,12 +73,12 @@ WLCSM_OBJS := $(foreach x, $(WLCSM_FILES
 
 OBJS += $(WLCSM_OBJS)
 STATIC_OBJS := $(filter-out wlcsm_lib_hspot.o, $(OBJS))
-ifneq ($(strip $(BUILD_BRCM_UNFWLCFG)),)
-conditional_build: $(WLCSM_EXT_BASE)/include/wlcsm_linux.h
-	ln -sf $(WLCSM_EXT_BASE)/include/wlcsm_linux.h  $(CURR_DIR)/include/wlcsm_linux.h
-	ln -sf $(WLCSM_EXT_BASE)/include/wl_common_defs.h  $(CURR_DIR)/include/wl_common_defs.h
-	@echo "skipping libwlcsm.so (not configured)"
-else
+#ifneq ($(strip $(BUILD_BRCM_UNFWLCFG)),)
+#conditional_build: $(WLCSM_EXT_BASE)/include/wlcsm_linux.h
+#	ln -sf $(WLCSM_EXT_BASE)/include/wlcsm_linux.h  $(CURR_DIR)/include/wlcsm_linux.h
+#	ln -sf $(WLCSM_EXT_BASE)/include/wl_common_defs.h  $(CURR_DIR)/include/wl_common_defs.h
+#	@echo "skipping libwlcsm.so (not configured)"
+#else
 conditional_build: $(LIBWLCSM_D)
 	ln -sf $(BUILD_DIR)/userspace/private/libs/wlcsm/include/wlcsm_defs.h  $(BUILD_DIR)/userspace/private/include/wlcsm_defs.h  
 	ln -sf $(BUILD_DIR)/userspace/private/libs/wlcsm/include/wlcsm_lib_api.h  $(BUILD_DIR)/userspace/private/include/wlcsm_lib_api.h  
@@ -98,7 +98,7 @@ endif
 	$(AR) $(ARFLAGS) $(LIBWLCSM_S) $(OBJS)
 	$(RANLIB) $(LIBWLCSM_S)
 	install -m 755 $(LIBWLCSM_D) $(INSTALL_DIR)/lib
-endif
+#endif
 
 install: all
 	install -m 755 $(LIBWLCSM_D) $(INSTALL_DIR)/lib
