--- a/cfe/cfe/board/bcm63xx_ram/src/bcm63xx_dtb.c
+++ b/cfe/cfe/board/bcm63xx_ram/src/bcm63xx_dtb.c
@@ -707,21 +707,22 @@ int dtb_set_bootargs(int dtb_id, char* b
     int nodeoffset, res = -1,proplen = 0;
     void* fdt = get_dtb_ptr(dtb_id);
 
-    char *node = NULL,*prop,**err_msg = &prop, *new_bootargs = NULL;
+    char node[sizeof(DT_CHOSEN_NODE)+sizeof(DT_ROOT_NODE)+1];
+    char *prop = DT_BOOTARGS_PROP;
+    char **err_msg = &prop;
+    char *new_bootargs = NULL;
+
     if (!fdt) {
         return res;
     }
-    node = KMALLOC(sizeof(DT_CHOSEN_NODE)+sizeof(DT_ROOT_NODE),0);
-    if (!node) {
-       goto err_out;
-    }
+
     sprintf(node,"%s%s",DT_ROOT_NODE,DT_CHOSEN_NODE);
+    node[sizeof(node)-1] = '\0';
     nodeoffset = fdt_path_offset(fdt, node);
     if (nodeoffset == -FDT_ERR_NOTFOUND) {
        *err_msg = node;
        goto err_out;
     }
-    prop = DT_BOOTARGS_PROP;
     propdata = fdt_getprop(fdt, nodeoffset, prop, &proplen);
 
     /* edit this to include mmc device root node in DT_BOOTARGS */
@@ -752,7 +753,6 @@ int dtb_set_bootargs(int dtb_id, char* b
 
     res = 0;
 err_out:
-    KFREE(node);
     if (bootargs) {
         KFREE(bootargs);
     }
