--- a/shared/opensource/boardparms/bcm963xx/boardparms_47622.c
+++ b/shared/opensource/boardparms/bcm963xx/boardparms_47622.c
@@ -819,6 +819,44 @@ static bp_elem_t g_bcm96755pathos[] = {
     {bp_last}
 };
 
+static bp_elem_t g_bcm96755augustus[] = {
+    {bp_cpBoardId,               .u.cp = "96755augustus"},
+    {bp_ulCompatChipId,          .u.ul = 0x6755},
+    /* use slower ddr clock in 6755 chip for now. May bump up to 1067MHz if test pass */
+    {bp_ulMemoryConfig,          .u.ul = BP_DDR_SPEED_800_11_11_11 | BP_DDR_TOTAL_SIZE_512MB| BP_DDR_DEVICE_WIDTH_16 | BP_DDR_TOTAL_WIDTH_16BIT | BP_DDR_SSC_CONFIG_1},
+    {bp_ulInterfaceEnable,       .u.ul = BP_PINMUX_FNTYPE_NAND},
+    {bp_ulInterfaceEnable,       .u.ul = BP_PINMUX_FNTYPE_HS_SPI},
+    {bp_usGphyBaseAddress,       .u.us = BCM947622_PHY_BASE},
+    {bp_ucPhyType0,              .u.uc = BP_ENET_NO_PHY},
+    {bp_usConfigType,            .u.us = BP_ENET_CONFIG_MMAP}, // Cross bar switch
+    {bp_ulPortMap,               .u.ul = 0x3},
+    {bp_ulPhyId0,                .u.ul = (BCM947622_PHY_BASE + 0x00) | (ADVERTISE_ALL_GMII | PHY_ADV_CFG_VALID)},
+    {bp_ulPhyId1,                .u.ul = 6 | PHY_INTEGRATED_VALID | MAC_IF_SERDES},
+    {bp_ulPhyId1,                .u.ul = 0x1f | PHY_EXTERNAL | PHY_TYPE_CL45GPHY},
+    {bp_ulPortFlags,             .u.ul = PORT_FLAG_SWAP_PAIR},
+    {bp_usSpiSlaveSelectNum,     .u.us = 0},
+    {bp_usSpiSlaveSelectGpioNum, .u.us = BP_GPIO_75_AL},
+    /* SPI SS1 must be configured but not be used for high speed greater than 30MHz */
+    {bp_usSpiSlaveSelectNum,     .u.us = 1},
+    {bp_usSpiSlaveSelectGpioNum, .u.us = BP_GPIO_76_AL},
+    {bp_ReservedAnyLed,          .u.us = BP_GPIO_62_AH},
+    /* HS_UART for dialog BT chip */
+    {bp_usIntfId,                .u.us = 1},
+    {bp_usIntfType,              .u.us = BP_INTF_TYPE_UART},
+    {bp_usPortNum,               .u.us = 1},
+    {bp_usUartSdin,              .u.us = BP_GPIO_5_AH},
+    {bp_usUartSdout,             .u.us = BP_GPIO_6_AH},
+    {bp_usIntfEnd},
+    /* I2C for Nuvoton MCU */
+    {bp_usIntfId,                .u.us = 3},
+    {bp_usIntfType,              .u.us = BP_INTF_TYPE_I2C},
+    {bp_usPortNum,               .u.us = 0},
+    {bp_usGpioI2cSda,            .u.us = BP_GPIO_16_AH},
+    {bp_usGpioI2cScl,            .u.us = BP_GPIO_17_AH},
+    {bp_usIntfEnd},
+    {bp_last}
+};
+
 bp_elem_t * g_BoardParms[] = {g_bcm947622sv, g_bcm947622sv_otp, g_bcm947622rfdvt, g_bcm947622rfdvt2, g_bcm947622ref1, g_bcm96755ref1, g_bcm947622eap, g_bcm947622eap2, g_bcm947622eaps, g_bcm947622sv_serdes, g_bcm947622ref1_sg, g_bcm96755ref1_sg, g_bcm96755tbrhx, \
-    g_bcm947622ref1_p250, g_bcm947622ref1_sg_p250, g_bcm96755ref1_p200, g_bcm96755ref1_sg_p200, g_bcm96755ref2, g_bcm947622leap, g_bcm96755ref1t_sg, g_bcm96755pathos, 0};
+    g_bcm947622ref1_p250, g_bcm947622ref1_sg_p250, g_bcm96755ref1_p200, g_bcm96755ref1_sg_p200, g_bcm96755ref2, g_bcm947622leap, g_bcm96755ref1t_sg, g_bcm96755pathos, g_bcm96755augustus, 0};
 
