--- a/shared/opensource/boardparms/bcm963xx/pinmux_table_47622.c
+++ b/shared/opensource/boardparms/bcm963xx/pinmux_table_47622.c
@@ -82,6 +82,10 @@ bp_pinmux_fn_defs_t g_pinmux_fn_defs[] =
 	{ BP_PINMUX_FNTYPE_xMII | 0, 68 | BP_PINMUX_VAL_1 },
 	{ BP_PINMUX_FNTYPE_xMII | 0, 69 | BP_PINMUX_VAL_1 },
 
+#if defined(CONFIG_RELEASE_UNUSED_RGMII_PIN)
+	{ BP_PINMUX_FNTYPE_xMII | 1, 68 | BP_PINMUX_VAL_1 },
+	{ BP_PINMUX_FNTYPE_xMII | 1, 69 | BP_PINMUX_VAL_1 },
+#else
 	{ BP_PINMUX_FNTYPE_xMII | 1, 56 | BP_PINMUX_VAL_1 | BP_PINMUX_PADCTL },
 	{ BP_PINMUX_FNTYPE_xMII | 1, 57 | BP_PINMUX_VAL_1 | BP_PINMUX_PADCTL },
 	{ BP_PINMUX_FNTYPE_xMII | 1, 58 | BP_PINMUX_VAL_1 | BP_PINMUX_PADCTL },
@@ -96,6 +100,7 @@ bp_pinmux_fn_defs_t g_pinmux_fn_defs[] =
 	{ BP_PINMUX_FNTYPE_xMII | 1, 67 | BP_PINMUX_VAL_1 | BP_PINMUX_PADCTL },
 	{ BP_PINMUX_FNTYPE_xMII | 1, 68 | BP_PINMUX_VAL_1 },
 	{ BP_PINMUX_FNTYPE_xMII | 1, 69 | BP_PINMUX_VAL_1 },
+#endif
 
 	{ BP_PINMUX_FNTYPE_NAND, 29 | BP_PINMUX_VAL_1 },
 	{ BP_PINMUX_FNTYPE_NAND, 40 | BP_PINMUX_VAL_1 },
--- a/shared/opensource/boardparms/bcm963xx/Makefile
+++ b/shared/opensource/boardparms/bcm963xx/Makefile
@@ -28,6 +28,10 @@ ifneq ($(CONFIG_BCM_ARCHER),)
 EXTRA_CFLAGS += -DCONFIG_BCM_SYSPORT
 endif
 
+ifeq ($(BRCM_BOARD_ID),"96755augustus")
+EXTRA_CFLAGS += -DCONFIG_RELEASE_UNUSED_RGMII_PIN
+endif
+
 -include $(TOPDIR)/Rules.make
 
 endif
