--- a/targets/config.in
+++ b/targets/config.in
@@ -2777,3 +2777,19 @@ fi
 
 endmenu
 
+mainmenu_option next_comment
+comment 'OpenSync Additions'
+
+bool "Opensync" BUILD_OPENSYNC
+bool "Plume target platform business logic" BUILD_PLUME
+if [ "$BUILD_OPENSYNC" = "y" ]; then
+    # technically uuid is required for miniupnd
+    define_bool BUILD_LIBUUID "y"
+    # curl binary is required to pull fw updates over ssl
+    define_bool BUILD_LIBCURL "y"
+    define_bool BUILD_LIBCURL_WITH_SSL "y"
+fi
+if [ "$BRCM_BOARD_ID" = "96755pathos" ]; then define_string TARGET "PATHOS_BCM52"; fi
+
+endmenu
+
