--- a/kernel/dts/47189/947189.dts
+++ b/kernel/dts/47189/947189.dts
@@ -17,6 +17,32 @@
 	aliases {
 	};
 
+#if defined (CONFIG_PSTORE_RAM)
+
+#if !CONFIG_PSTORE_RAM_ADDRESS
+/* Default PSTORE address for this platform */
+/* Use last 1.25M of 128 MB RAM */
+#define CONFIG_PSTORE_RAM_ADDRESS_TAG   07EC0000
+#define CONFIG_PSTORE_RAM_ADDRESS     0x07EC0000
+#endif
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		/* Reserve the region for ramoops, use 1.25M at CONFIG_PSTORE_RAM_ADDRESS */
+		ramoops@CONFIG_PSTORE_RAM_ADDRESS_TAG {
+			compatible = "ramoops";
+			reg = <CONFIG_PSTORE_RAM_ADDRESS 0x140000>;
+			record-size = <0x4000>;
+			console-size = <0x4000>;
+			pmsg-size = <0x10000>;
+			ecc-size = <1>;
+		};
+	};
+#endif
+
 	memory {
 		device_type = "memory";
 		reg = <0x0 0x8000000>;
--- a/kernel/dts/47622/947622.dts
+++ b/kernel/dts/47622/947622.dts
@@ -132,6 +132,32 @@
 	};
 #endif
 
+#if defined (CONFIG_PSTORE_RAM)
+
+#if !CONFIG_PSTORE_RAM_ADDRESS
+/* Default PSTORE address for this platform */
+/* Use last 1.25M of 512 MB RAM */
+#define CONFIG_PSTORE_RAM_ADDRESS_TAG   1FEC0000
+#define CONFIG_PSTORE_RAM_ADDRESS     0x1FEC0000
+#endif
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		/* Reserve the region for ramoops, use 1.25M at CONFIG_PSTORE_RAM_ADDRESS */
+		ramoops@CONFIG_PSTORE_RAM_ADDRESS_TAG {
+			compatible = "ramoops";
+			reg = <CONFIG_PSTORE_RAM_ADDRESS 0x140000>;
+			record-size = <0x4000>;
+			console-size = <0x4000>;
+			pmsg-size = <0x10000>;
+			ecc-size = <1>;
+		};
+	};
+#endif
+
 #if defined (CONFIG_OPTEE)
 	psci {
 		compatible	= "arm,psci-0.2";
--- a/kernel/dts/gen_dt_bindings
+++ b/kernel/dts/gen_dt_bindings
@@ -69,6 +69,17 @@ then
 	echo "#define CONFIG_LEDS_KTD206X ${CONFIG_LEDS_KTD206X}" >> ${2}
 fi
 
+if [ "${CONFIG_PSTORE_RAM}" == "m" -o "${CONFIG_PSTORE_RAM}" == "y" ]
+then
+	echo "#define CONFIG_PSTORE_RAM ${CONFIG_PSTORE_RAM}" >> ${2}
+fi
+
+if [ -n "${CONFIG_PSTORE_RAM_ADDRESS}" ]
+then
+	echo "#define CONFIG_PSTORE_RAM_ADDRESS ${CONFIG_PSTORE_RAM_ADDRESS}" >> ${2}
+	echo "#define CONFIG_PSTORE_RAM_ADDRESS_TAG ${CONFIG_PSTORE_RAM_ADDRESS#0x}" >> ${2}
+fi
+
 # Support string defines like:
 # #define SOMETHING "something"
 # The input here will be ' #define "SOMETHING" "something" ' which means
--- a/kernel/linux-4.1/Kconfig.bcm
+++ b/kernel/linux-4.1/Kconfig.bcm
@@ -1875,6 +1875,13 @@ config OVS_SKIP_ACCEL_ACTION_HOOK
         depends on OVS_SKIP_ACCEL_ACTION
         default n
 
+config PSTORE_RAM_ADDRESS
+	hex "PSTORE RAM address"
+	depends on PSTORE_RAM
+	default 0
+	help
+	  If 0 use platform default
+
 menu "Autodetected Drivers"
 source "../../bcmdrivers/Kconfig.autogen"
 endmenu
