--- a/userspace/public/apps/dropbear/Makefile
+++ b/userspace/public/apps/dropbear/Makefile
@@ -74,6 +74,9 @@ check_untar_patch_configure: sanity_chec
 	    if [ "x$(BUILD_BRCM_CMS)" != "x" ]; then \
 		patch -p1 -b -s -d$(APP) < $(APP)_cms.patch; \
 	    fi; \
+	    if [ "x$(BUILD_OPENSYNC)" != "x" ]; then \
+		mkdir -p $(APP)/patches && cd $(APP)/patches && for i in ../../???-*.patch; do ln -vsf $$i .; done && ls > .series && mv -v .series series && cd .. && quilt push -a || exit $$?; \
+	    fi; \
 	    (cd $(APP); ./configure CFLAGS="$(CFLAGS)" LDFLAGS="$(BCM_BLT32_FLAGS) -L$(INSTALL_DIR)/lib" LIBS="$(LIBS)" --host=$(TOOLCHAIN_PREFIX) \
         --target=$(TOOLCHAIN_PREFIX) \
         --disable-largefile --disable-zlib  --disable-openpty --disable-syslog  --disable-shadow  --disable-lastlog  --disable-utmp --disable-utmpx --disable-wtmp --disable-wtmpx  \
