
+++ //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/47622/947622.dts
--- //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/47622/947622.dts

@@ -42,7 +42,7 @@
 	#size-cells = <1>;
 
 	chosen {
-		bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0";
+		bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0 pci=pcie_bus_safe";
 	}; 
 
 	cpus {
@@ -296,4 +296,4 @@
 		compatible = "brcm,therm";
 		status = "okay";
 	};
-};
\ No newline at end of file
+};

+++ //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/63178/963178.dts
--- //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/63178/963178.dts

@@ -42,7 +42,7 @@
 	#size-cells = <1>;
 
 	chosen {
-		bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0";
+		bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0 pci=pcie_bus_safe";
 	}; 
 
 	cpus {

+++ //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/6878/96878.dts
--- //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/6878/96878.dts

@@ -31,7 +31,7 @@
     #size-cells = <1>;
    
     chosen {
-        bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0";
+        bootargs = "console=ttyAMA0 earlyprintk debug irqaffinity=0 pci=pcie_bus_safe";
     };
 
 	cpus {
