
+++ //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/47622/947622.dts
--- //depot/CommEngine/patches/WLAN/PATCH_502L07_P1/kernel/dts/47622/947622.dts

@@ -132,9 +132,13 @@
 	pmu {
 		compatible = "arm,cortex-a7-pmu";
 		interrupts = <GIC_SPI PMU_CORE0_SPI IRQ_TYPE_LEVEL_HIGH>,
-			     <GIC_SPI PMU_CORE1_SPI IRQ_TYPE_LEVEL_HIGH>;
+			     <GIC_SPI PMU_CORE1_SPI IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI PMU_CORE2_SPI IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI PMU_CORE3_SPI IRQ_TYPE_LEVEL_HIGH>;
 		interrupt-affinity = <&CA7_0>,
-				     <&CA7_1>;
+				     <&CA7_1>,
+				     <&CA7_2>,
+				     <&CA7_3>;
 	};
 	
 	uartclk: uartclk {
