From 951c117226d0517ca9ca9d657b47da68d75711c0 Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Tue, 21 Jan 2020 15:48:18 -0800
Subject: [PATCH] OWRT-4719 get_dev_stats added to b53

---
 drivers/net/dsa/b53/b53_priv.h   |   6 +-

diff --git a/drivers/net/dsa/b53/b53_priv.h b/drivers/net/dsa/b53/b53_priv.h
index ce1880e..7d4b1f0 100644
--- a/drivers/net/dsa/b53/b53_priv.h
+++ b/drivers/net/dsa/b53/b53_priv.h
@@ -118,6 +118,7 @@ struct b53_device {
 	struct b53_vlan *vlans;
 	unsigned int num_ports;
 	struct b53_port *ports;
+	struct rtnl_link_stats64 stats[B53_N_PORTS];
 };
 
 #define b53_for_each_port(dev, i) \
@@ -201,10 +202,7 @@ int b53_switch_detect(struct b53_device *dev);
 
 int b53_switch_register(struct b53_device *dev);
 
-static inline void b53_switch_remove(struct b53_device *dev)
-{
-	dsa_unregister_switch(dev->ds);
-}
+void b53_switch_remove(struct b53_device *dev);
 
 static inline int b53_read8(struct b53_device *dev, u8 page, u8 reg, u8 *val)
 {
-- 
1.9.1

