diff -urN a/platform/cfg80211/src/lib/target/wiphy_info.c b/platform/cfg80211/src/lib/target/wiphy_info.c
--- a/platform/cfg80211/src/lib/target/wiphy_info.c	2021-03-05 02:29:37.639000882 +0530
+++ b/platform/cfg80211/src/lib/target/wiphy_info.c	2021-02-22 11:36:33.131646304 +0530
@@ -77,6 +77,7 @@
     { 0, 0, "qca,wifi-ar956x", "qca9563", "Dragonfly" },
     { 0, 0, "qcom,cnss-qca8074", "qca8074", "Hawkeye" },
     { 0, 0, "qcom,cnss-qca8074v2", "qca8074", "Hawkeye" },
+    { 0xaa52, 0x14e4, 0, "bcm43602", "Northstar" }, /* sr400 BCM53xx details*/
 };
 
 /* runtime data */

