diff -bru htpdate-1.0.4.orig/htpdate.c htpdate-1.0.4/htpdate.c
--- htpdate-1.0.4.orig/htpdate.c	2019-02-20 13:58:49.783633143 +0100
+++ htpdate-1.0.4/htpdate.c	2019-02-20 13:59:05.556019283 +0100
@@ -1067,10 +1067,12 @@
 			exit(1);
 	}
 
+#if 0
 	/* After first poll cycle do not step through time, only adjust */
 	if ( setmode != 3 ) {
 		setmode = 1;
 	}
+#endif
 
 	} while ( daemonize_dbg );		/* end of infinite while loop */
 
