--- a/default_options.h
+++ b/default_options.h
@@ -13,6 +13,11 @@ Options can also be defined with -DDROPB
 
 IMPORTANT: Some options will require "make clean" after changes */
 
+#if !defined(DROPBEAR_CLIENT) && !defined(DROPBEAR_SERVER)
+#define DROPBEAR_SERVER 1
+#define DROPBEAR_CLIENT 1
+#endif
+
 #define DROPBEAR_DEFPORT "22"
 
 /* Listen on all interfaces */
@@ -34,7 +39,7 @@ IMPORTANT: Some options will require "ma
  * Both of these flags can be defined at once, don't compile without at least
  * one of them. */
 #define NON_INETD_MODE 1
-#define INETD_MODE 1
+#define INETD_MODE 0
 
 /* Include verbose debug output, enabled with -v at runtime. 
  * This will add a reasonable amount to your executable size. */
@@ -71,7 +76,7 @@ IMPORTANT: Some options will require "ma
 
 /* Enable "Netcat mode" option. This will forward standard input/output
  * to a remote TCP-forwarded connection */
-#define DROPBEAR_CLI_NETCAT 1
+#define DROPBEAR_CLI_NETCAT 0
 
 /* Whether to support "-c" and "-m" flags to choose ciphers/MACs at runtime */
 #define DROPBEAR_USER_ALGO_LIST 1
@@ -82,7 +87,7 @@ IMPORTANT: Some options will require "ma
  * Including both AES keysize variants (128 and 256) will result in 
  * a minimal size increase */
 #define DROPBEAR_AES128 1
-#define DROPBEAR_3DES 1
+#define DROPBEAR_3DES 0
 #define DROPBEAR_AES256 1
 #define DROPBEAR_TWOFISH256 0
 #define DROPBEAR_TWOFISH128 0
@@ -91,7 +96,7 @@ IMPORTANT: Some options will require "ma
 
 /* Enable CBC mode for ciphers. This has security issues though
  * is the most compatible with older SSH implementations */
-#define DROPBEAR_ENABLE_CBC_MODE 1
+#define DROPBEAR_ENABLE_CBC_MODE 0
 
 /* Enable "Counter Mode" for ciphers. This is more secure than
  * CBC mode against certain attacks. It is recommended for security
@@ -101,7 +106,7 @@ IMPORTANT: Some options will require "ma
 /* Message integrity. sha2-256 is recommended as a default, 
    sha1 for compatibility */
 #define DROPBEAR_SHA1_HMAC 1
-#define DROPBEAR_SHA1_96_HMAC 1
+#define DROPBEAR_SHA1_96_HMAC 0
 #define DROPBEAR_SHA2_256_HMAC 1
 
 /* Hostkey/public key algorithms - at least one required, these are used
@@ -111,7 +116,7 @@ IMPORTANT: Some options will require "ma
  * DSS may be necessary to connect to some systems though
    is not recommended for new keys */
 #define DROPBEAR_RSA 1
-#define DROPBEAR_DSS 1
+#define DROPBEAR_DSS 0
 /* ECDSA is significantly faster than RSA or DSS. Compiling in ECC
  * code (either ECDSA or ECDH) increases binary size - around 30kB
  * on x86-64 */
