--- a/cli-kex.c
+++ b/cli-kex.c
@@ -302,8 +302,14 @@ static void checkhostkey(const unsigned
 
 	hostsfile = open_known_hosts_file(&readonly);
 	if (!hostsfile)	{
-		ask_to_confirm(keyblob, keybloblen, algoname);
-		/* ask_to_confirm will exit upon failure */
+        if (!cli_opts.dropunknown) {
+		    ask_to_confirm(keyblob, keybloblen, algoname);
+		    /* ask_to_confirm will exit upon failure */
+        }
+        else {
+            /* drop connection in this case as there are no known hosts */
+	        dropbear_exit("Host key required, validation failed");
+        }
 		return;
 	}
 	
@@ -372,7 +378,12 @@ static void checkhostkey(const unsigned
 	} while (1); /* keep going 'til something happens */
 
 	/* Key doesn't exist yet */
-	ask_to_confirm(keyblob, keybloblen, algoname);
+    if (!cli_opts.dropunknown) {
+	    ask_to_confirm(keyblob, keybloblen, algoname);
+    }
+    else {
+        dropbear_exit("Host key required, validation failed");
+    }
 
 	/* If we get here, they said yes */
 
--- a/cli-runopts.c
+++ b/cli-runopts.c
@@ -66,6 +66,7 @@ static void printhelp() {
 					"-y -y Don't perform any remote host key checking (caution)\n"
 					"-s    Request a subsystem (use by external sftp)\n"
 					"-o option     Set option in OpenSSH-like format ('-o help' to list options)\n"
+                    "-D    drop connection if host unknown\n"
 #if DROPBEAR_CLI_PUBKEY_AUTH
 					"-i <identityfile>   (multiple allowed, default %s)\n"
 #endif
@@ -146,6 +147,7 @@ void cli_getopts(int argc, char ** argv)
 	cli_opts.always_accept_key = 0;
 	cli_opts.no_hostkey_check = 0;
 	cli_opts.is_subsystem = 0;
+	cli_opts.dropunknown = 0;
 #if DROPBEAR_CLI_PUBKEY_AUTH
 	cli_opts.privkeys = list_new();
 #endif
@@ -233,6 +235,9 @@ void cli_getopts(int argc, char ** argv)
 				case 'f':
 					cli_opts.backgrounded = 1;
 					break;
+				case 'D':
+					cli_opts.dropunknown = 1;
+					break;
 				case 's':
 					cli_opts.is_subsystem = 1;
 					break;
@@ -307,7 +312,6 @@ void cli_getopts(int argc, char ** argv)
 				case 'c':
 				case 'm':
 #endif
-				case 'D':
 #if !DROPBEAR_CLI_REMOTETCPFWD
 				case 'R':
 #endif
--- a/runopts.h
+++ b/runopts.h
@@ -148,6 +148,7 @@ typedef struct cli_runopts {
 	int no_cmd;
 	int backgrounded;
 	int is_subsystem;
+    int dropunknown;
 #if DROPBEAR_CLI_PUBKEY_AUTH
 	m_list *privkeys; /* Keys to use for public-key auth */
 #endif
