--- a/userspace/gpl/apps/ppp/Makefile
+++ b/userspace/gpl/apps/ppp/Makefile
@@ -34,7 +34,7 @@ clean:
 
 #.PHONY: check_untar_patch_configure conditional_build
 
-ifneq ($(strip $(BUILD_ACCEL_PPTP)),)
+ifneq ($(if $(BUILD_OPENSYNC),,$(strip $(BUILD_ACCEL_PPTP))),)
 
 check_untar_patch_configure: sanity_check
 	@if [ ! -e $(APPDIR)/Makefile ]; then \
--- a/userspace/public/apps/pppd/autodetect
+++ b/userspace/public/apps/pppd/autodetect
@@ -1,4 +1,4 @@
-condition: ifneq ($(or (strip $(BUILD_BRCM_CMS)),$(BUILD_BRCM_HNDROUTER)),)
+condition: ifneq ($(if $(BUILD_OPENSYNC),,$(or $(strip $(BUILD_BRCM_CMS)),$(BUILD_BRCM_HNDROUTER))),)
 dependson: public/libs/cms_msg public/libs/cms_util public/libs/bcm_flashutil public/libs/bcm_boardctl
 
 
--- a/userspace/public/apps/pppd/Makefile
+++ b/userspace/public/apps/pppd/Makefile
@@ -11,7 +11,7 @@ export CMS_COMMON_LIBS
 export BCM_LD_FLAGS
 export DESKTOP_LINUX
 
-ifneq ($(strip $(BUILD_PPPD)),)
+ifneq ($(if $(BUILD_OPENSYNC),,$(strip $(BUILD_PPPD))),)
 conditional_build: all
 else
 conditional_build:
@@ -20,7 +20,7 @@ endif
 
 
 all: 
-ifneq ($(strip $(BUILD_BRCM_CMS)),)
+ifneq ($(if $(BUILD_OPENSYNC),,$(strip $(BUILD_BRCM_CMS))),)
 	$(MAKE) -C ppp/pppoe $(BUILD_PPPD)
 else
 	$(MAKE) -C ppp/pppoe
--- a/targets/fs.src/etc/ppp/options
+++ /dev/null
@@ -1,3 +0,0 @@
-lock
-plugin /usr/sbin/pppoe.so
-name test
