Index: build-cyrus-opensync-y/userspace/gpl/apps/iptables/Makefile
===================================================================
--- build-cyrus-opensync-y.orig/userspace/gpl/apps/iptables/Makefile
+++ build-cyrus-opensync-y/userspace/gpl/apps/iptables/Makefile
@@ -32,7 +32,7 @@ ifneq ($(strip $(BRCM_KERNEL_NF_NAT)),)
 PF4_EXT+=DNAT MASQUERADE REDIRECT SNAT
 endif
 ifneq ($(strip $(BRCM_KERNEL_NF_MANGLE)),)
-PFX_EXT+=dscp DSCP tos TOS mac mark MARK
+PFX_EXT+=dscp DSCP tos TOS mac mark MARK TPROXY
 endif
 ifneq ($(strip $(BUILD_IPV6)),)
 PF6_EXT+=ah dst eui64 frag hbh hl icmp6 ipv6header mh rt HL LOG REJECT
