--- a/userspace/public/apps/openvswitch/Makefile
+++ b/userspace/public/apps/openvswitch/Makefile
@@ -91,6 +91,7 @@ inner_kernelmodulebuild:
 	cd $(APP)/datapath/linux; $(MAKE)
 	install -m 777 -d $(INSTALL_DIR)/opt/modules/openvswitch
 	install -m 644 ./$(APP)/datapath/linux/openvswitch.ko  $(INSTALL_DIR)/opt/modules/openvswitch/openvswitch.ko 	
+	$(STRIP) -g $(INSTALL_DIR)/opt/modules/openvswitch/openvswitch.ko
 		
 userspace_build:
 	    (cd $(APP); unset CFLAGS; \
