--- a/bcmdrivers/opensource/net/enet/impl7/Makefile
+++ b/bcmdrivers/opensource/net/enet/impl7/Makefile
@@ -130,12 +130,12 @@ endif
 # Except 47189 and HNDROUTER, all BCACPE platforms
 ifneq ($(strip $(BUILD_BRCM_HNDROUTER)),)
 else
-ifeq (,$(filter $(BRCM_CHIP),47189))
+ifeq (,$(filter $(BRCM_CHIP),47189 47622))
 EXTRA_CFLAGS += -DSEPARATE_MAC_FOR_WAN_INTERFACES
 endif
 endif
 
-ifneq (,$(filter $(BRCM_CHIP),47189))
+ifneq (,$(filter $(BRCM_CHIP),47189 47622))
 EXTRA_CFLAGS += -DSEPARATE_MAC_FOR_LAN_INTERFACES
 endif
 
