diff --git a/scripts/feeds b/scripts/feeds
index eee0a50717..50c2f5ea41 100755
--- a/scripts/feeds
+++ b/scripts/feeds
@@ -536,12 +536,7 @@ sub install_src {
 	# If it's a core package and we don't want to override, just return
 	my $override = 0;
 	if (is_core_src($name)) {
-		if (!$force) {
-			if ($name ne "toolchain" && $name ne "linux") {
-				warn "WARNING: Not overriding core package '$name'; use -f to force\n";
-			}
-			return 0;
-		}
+		return 0 unless $force;
 		$override = 1;
 	}
 
